### PR TITLE
feat: prettier trait

### PR DESCRIPTION
**Pull request by yuanbohan**
Ref: https://github.com/prometheus/prometheus/blob/main/promql/parser/prettier.go
Test: https://github.com/prometheus/prometheus/blob/main/promql/parser/prettier_test.go
close: #55

### CODE REVIEW

File: src/parser/mod.rs [hunk 1]
```diff
@@ -42,3 +42,52 @@ pub use value::{Value, ValueType};
...
+/// The level of a Node determines if it should be indented or not.
+/// The answer to the 1 is NO if the level passed is 0. This means, the
+/// parent Node did not apply a new line, so the current Node must not
+/// apply any indentation as prefix.
+/// If level > 1, a new line is applied by the parent. So, the current Node
+/// should prefix an indentation before writing any of its content. This indentation
+/// will be ([level/depth of current Node] * "  ").
+///
+/// The answer to 2 is YES if the normalized length of the current Node exceeds
+/// the maxCharactersPerLine limit. Hence, it applies the indentation equal to
```
Comment (evenyag) [inline src/parser/mod.rs:68]:
```suggestion
/// the [MAX_CHARACTERS_PER_LINE] limit. Hence, it applies the indentation equal to
```

File: src/parser/ast.rs [hunk 2]
```diff
@@ -110,21 +112,29 @@ pub struct BinModifier {
...
+        let mut s = String::from(self.bool_str());
+
         if let Some(matching) = &self.matching {
             match matching {
-                LabelModifier::Include(ls) => write!(f, "on ({ls}) ")?,
-                LabelModifier::Exclude(ls) if !ls.is_empty() => write!(f, "ignoring ({ls}) ")?,
+                LabelModifier::Include(ls) => s.push_str(&format!("on ({ls}) ")),
+                LabelModifier::Exclude(ls) if !ls.is_empty() => {
+                    s.push_str(&format!("ignoring ({ls}) "))
+                }
```
[the remark this answers is not in the document]
Comment (yuanbohan) [reply src/parser/ast.rs]: done

**Review by evenyag — DISMISSED** [review]
LGTM

**Review by waynexia — APPROVED** [review]
I skimmed the tests from both sides and they look identical 👍

File: src/parser/mod.rs [hunk 3]
```diff
@@ -42,3 +42,52 @@ pub use value::{Value, ValueType};
...
+/// If level > 1, a new line is applied by the parent. So, the current Node
+/// should prefix an indentation before writing any of its content. This indentation
+/// will be ([level/depth of current Node] * "  ").
+///
+/// The answer to 2 is YES if the normalized length of the current Node exceeds
+/// the [MAX_CHARACTERS_PER_LINE] limit. Hence, it applies the indentation equal to
+/// its depth and increments the level by 1 before passing down the child.
+/// If the answer is NO, the current Node returns the normalized string value of itself.
+pub trait Prettier: std::fmt::Display {
+    fn pretty(&self, level: usize, max: usize) -> String {
```
[the remark this answers is not in the document]
Comment (yuanbohan) [reply src/parser/mod.rs:72]: Yes, but the name is too long, I will leave a comment.

File: src/parser/ast.rs [hunk 4]
```diff
@@ -1653,7 +1810,612 @@ mod tests {
...
+        }
+    }
+
+    #[test]
+    fn test_binary_expr_pretty() {
+        let cases = vec![
+            ("a+b", "a + b"),
+            (
+                "a == bool 1",
+                "  a
```
[the remark this answers is not in the document]
Comment (yuanbohan) [reply src/parser/ast.rs:1918]: As for BinaryExpr, the Op is at the front, with LHS and RHS with 2 indents.

**Review by evenyag — APPROVED** [review]

**Pull request merged**